### PR TITLE
feat: siopv2

### PR DESCRIPTION
**Pull request by Ad96el**

### CODE REVIEW

File: login-frontend/src/App.tsx [hunk 1]
```diff
@@ -28,68 +32,134 @@ export function App() {
...
+      const extension = new FormData(form).get('extension') as string;
+      const Dids = await kilt[extension].getDidList();
+
+      let did = Dids[0].did;
+
+      const didDocument = await Did.resolve(did);
+
+      // jwt parts
+      const keyURI = didDocument?.document?.authentication[0].id.replace("#0x", "");
+      const header = { alg: "EdDSA", typ: "JWT", keyURI }
```
Comment (trusch) [inline login-frontend/src/App.tsx:103]:
```suggestion
      const header = { alg: "EdDSA", typ: "JWT", kid: keyURI }
```

File: login-frontend/src/App.tsx [hunk 2]
```diff
@@ -28,68 +32,134 @@ export function App() {
...
+      const body = { iss: did, sub: did, nonce };
```
Comment (trusch) [inline login-frontend/src/App.tsx:104]: exp, nbf, ...

File: login-frontend/src/App.tsx [hunk 3]
```diff
@@ -87,32 +85,30 @@ export function App() {
 
   const handleSIOPV2Login = useCallback(
     async (nonce: string, event: FormEvent<HTMLFormElement>) => {
-
-      connect(process.env.DATABASE_URL ? process.env.DATABASE_URL : "wss://peregrine.kilt.io")
+      connect(process.env.DATABASE_URL ? process.env.DATABASE_URL : 'wss://peregrine.kilt.io');
```
Comment (trusch) [inline login-frontend/src/App.tsx:88]: Strange env Name for a fullnode

Comment (Ad96el) [reply login-frontend/src/App.tsx]: [3748bb2](https://github.com/KILTprotocol/opendid/pull/28/commits/3748bb284a53dc5dc989ae130bf5bb42adbf01a6)

File: login-frontend/src/App.tsx (hunk 2 above)
Comment (Ad96el) [reply login-frontend/src/App.tsx]: [ebb1fdd](https://github.com/KILTprotocol/opendid/pull/28/commits/ebb1fddb4c05cc0de39eb3148a1c614150699622)

File: login-frontend/src/App.tsx (hunk 1 above)
Comment (Ad96el) [reply login-frontend/src/App.tsx]: [ebb1fdd](https://github.com/KILTprotocol/opendid/pull/28/commits/ebb1fddb4c05cc0de39eb3148a1c614150699622)

File: login-frontend/src/App.tsx [hunk 4]
```diff
@@ -28,68 +30,136 @@ export function App() {
...
+      const didLoginResponse = await fetch(url, {
+        method: 'POST',
+      });
+
+      if (didLoginResponse.status >= 400) {
+        const credentialResponseData = await didLoginResponse.text();
+        throw new Error(credentialResponseData);
+      }
+
+      if (didLoginResponse.status === 204) {
```
Comment (trusch) [inline login-frontend/src/App.tsx:128]: all http status codes < 204 are not handled, you could throw an error

File: src/routes/did.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    let oidc_context = session
+        .get::<AuthorizeQueryParameters>(OIDC_SESSION_KEY)
+        .map_err(|_| Error::OauthNoSession)?
+        .ok_or(Error::OauthInvalidClientId)?;
+
+    let nonce = oidc_context.nonce;
+
+    let mut hasher =  Sha512::new();
+
+    let parts : Vec<&str> = jwt_token.split('.').collect();
```
Comment (trusch) [inline src/routes/did.rs:62]: Bette check if you really have three parts, otherwise the [2] could panic

File: src/routes/did.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        } else {
+            false
+        }
+    };
+
+    if !valid {
+        return Err(Error::InvalidDidSignature);
+    }
+
+    let w3n = kilt::get_w3n(&jwt_payload.iss, &cli).await.unwrap_or("".into());
```
Comment (trusch) [inline src/routes/did.rs:142]: We need to check if payload.iss is actually the DID that signed the JWT. Otherwise I could just put your DID there

**Review by trusch — CHANGES_REQUESTED** [review]
One or two minor things :) Only real bug is not checking payload.iss for being the same as sender and also being the same as payload.sub

File: src/routes/did.rs (hunk 5 above)
Comment (Ad96el) [reply src/routes/did.rs]: Always forget safe array operations :face_exhaling: [e289e12](https://github.com/KILTprotocol/opendid/pull/28/commits/e289e129e41c0e4639110fb340b9ae8736edb8c4)

File: login-frontend/src/App.tsx (hunk 4 above)
Comment (Ad96el) [reply login-frontend/src/App.tsx:127]: updated in: [4b81cda](https://github.com/KILTprotocol/opendid/pull/28/commits/4b81cda5a544f6c326a61488174c5ae147bf3fc3)

File: src/routes/did.rs (hunk 6 above)
Comment (Ad96el) [reply src/routes/did.rs]: I do but looks like it was not declarative enough. Is this better? [aa52209](https://github.com/KILTprotocol/opendid/pull/28/commits/aa52209ad8f05798dacd7ebdc852cf69777cc910)

File: src/routes/did.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        let app_state = app_state.read()?;
+        app_state.kilt_endpoint.clone()
+    };
+    let cli = kilt::connect(&endpoint)
+        .await
+        .map_err(|_| Error::CantConnectToBlockchain)?;
+
+    let did_document = get_did_doc(sender, &cli).await?;
+
+    let signed_key = hex_decode(jwt_header.kid.trim_start_matches("0x"))
```
Comment (trusch) [inline src/routes/did.rs:124]: I didn't catch that in the first review: The kid needs to be the full kid of the key used to sign the payload. Therefore it must contain the DID that signed it. 
It has to look like `kid: "did:kilt:4abcdefg#0xdeadbeaf112233"`. 
In the current code you just get the DID doc of the DID which is listed as sub and iss, and check that the key in kid is in that doc, so this is fine, but still not really compliant

Comment (Ad96el) [reply src/routes/did.rs]: updated in: [bb1fb58](https://github.com/KILTprotocol/opendid/pull/28/commits/bb1fb58e552c2bfba19d7c7e8f0e27d0e0e4957b)

**Review by trusch — APPROVED** [review]

**Pull request merged**